util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,641 @@
+		fs:      fs,
+		dirname: dirname,
+		cache:   cache,
+	}
+	pool.mu.readers = make(map[int]struct {
+		reader *sstable.Reader
+		refs   *atomic.Int32
+	})
+	return pool
+}
+
+func (p *sstReaderPool) get(fileNum int) (*sstable.Reader, error) {
+	var reader *sstable.Reader
+	withLock(p.mu.RLocker(), func() {
+		if v, ok := p.mu.readers[fileNum]; ok {
+			v.refs.Add(1)
+			reader = v.reader
+		}
+	})
+	if reader != nil {
+		return reader, nil
+	}
+
+	// Create a new reader. Note that we don't hold the lock here,
+	// since creating a reader can be expensive, and we don't want
+	// to block other readers' creation.
+	f, err := p.fs.Open(makeFilename(p.fs, p.dirname, fileNum))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	reader, err = sstable.NewReader(f, sstable.ReaderOptions{
+		Cache: p.cache,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var toRet, toClose *sstable.Reader
+	withLock(&p.mu, func() {
+		// Someone might have added the reader to the pool.
+		if v, ok := p.mu.readers[fileNum]; ok {
+			toClose = reader
+			v.refs.Add(1)
+			toRet = v.reader
+			return
+		}
+		refs := new(atomic.Int32)
+		refs.Store(1)
+		p.mu.readers[fileNum] = struct {
+			reader *sstable.Reader
+			refs   *atomic.Int32
+		}{reader: reader, refs: refs}
+		toRet = reader
+	})
+	if toClose != nil {
+		if err := toClose.Close(); err != nil {
+			_ = p.unref(fileNum)
+			return nil, errors.Trace(err)
+		}
+	}
+	return toRet, nil
+}
+
+func (p *sstReaderPool) unref(fileNum int) error {
+	var closer io.Closer
+	withLock(&p.mu, func() {
+		v, ok := p.mu.readers[fileNum]
+		if !ok {
+			panic(fmt.Sprintf("sstReaderPool: unref a reader that does not exist: %d", fileNum))
+		}
+		if v.refs.Add(-1) == 0 {
+			delete(p.mu.readers, fileNum)
+			closer = v.reader
+		}
+	})
+	if closer != nil {
+		return errors.Trace(closer.Close())
+	}
+	return nil
+}
+
+func withLock(l sync.Locker, f func()) {
+	l.Lock()
+	defer l.Unlock()
+	f()
+}
+
+type sstIter struct {
+	iter    sstable.Iterator
+	key     *sstable.InternalKey
+	value   []byte
+	onClose func() error
+}
+
+func (si *sstIter) Seek(key []byte) bool {
+	si.key, si.value = si.iter.SeekGE(key)
+	return si.key != nil
+}
+
+func (si *sstIter) First() bool {
+	si.key, si.value = si.iter.SeekGE(nil)
+	return si.key != nil
+}
+
+func (si *sstIter) Next() bool {
+	si.key, si.value = si.iter.Next()
+	return si.key != nil
+}
+
+func (si *sstIter) Last() bool {
+	si.key, si.value = si.iter.Last()
+	return si.key != nil
+}
+
+func (si *sstIter) Valid() bool {
+	return si.key != nil
+}
+
+func (si *sstIter) Error() error {
+	return errors.Trace(si.iter.Error())
+}
+
+func (si *sstIter) UnsafeKey() []byte {
+	return si.key.UserKey
+}
+
+func (si *sstIter) UnsafeValue() []byte {
+	return si.value
+}
+
+func (si *sstIter) Close() error {
+	err := si.iter.Close()
+	if si.onClose != nil {
+		err = goerrors.Join(err, si.onClose())
+	}
+	return errors.Trace(err)
+}
+
+type mergingIter struct {
+	heap mergingIterHeap
```

Comment:
>  if a file's range contains heap[0].key

No, it just means the iterator has been opened. It may be closed because of there is no real key after `heap[0].key`.